### PR TITLE
Coproduct improvements

### PR DESCRIPTION
**Pull request by ghostbuster91**
This references #84 

This PR is mainly about making coproduct reference-able.
Once I added object info to SCoproduct we also gained a benefit of being able to differentiate between generic coproducts ;)
I also figured out that we don't need a separate step to replace fileds within objects with references as it can be done directly in `TSchemaToOSchema`.

Btw: I don't like this `objectable` name, so feel free to drop any suggestions :)

### CODE REVIEW

File: core/src/main/scala/tapir/Schema.scala [hunk 1]
```diff
@@ -17,7 +17,7 @@ object Schema {
   case object SBoolean extends Schema {
     def show: String = "boolean"
   }
-  case class SObject(info: SObjectInfo, fields: Iterable[(String, Schema)], required: Iterable[String]) extends Schema {
+  case class SObject(info: SObjectInfo, fields: Iterable[(String, Schema)], required: Iterable[String]) extends SObjectable {
```
Comment (adamw) [inline core/src/main/scala/tapir/Schema.scala]: maybe rename:
* `SObject` to `SProductObject`
* `SCoproduct` to `SCoproductObject`
* `SObjectable` to `SObject`?

Comment (adamw) [reply core/src/main/scala/tapir/Schema.scala]: Or maybe even simpler, `SRecord`, `SSum` and `SObject` respectively?

Comment (adamw) [reply core/src/main/scala/tapir/Schema.scala]: I think Record/Sum might be easier to understand :)

File: docs/openapi-docs/src/test/resources/expected_coproduct.yml [hunk 2]
```diff
@@ -12,11 +12,13 @@ paths:
...
-                oneOf:
-                  - $ref: '#/components/schemas/Person'
-                  - $ref: '#/components/schemas/Organization'
+                $ref: '#/components/schemas/Entity'
 components:
   schemas:
+    Entity:
+      oneOf:
+        - $ref: '#/components/schemas/Person'
+        - $ref: '#/components/schemas/Organization'
```
Comment (adamw) [inline docs/openapi-docs/src/test/resources/expected_coproduct.yml]: that's valid for openapi?

Comment (ghostbuster91) [reply docs/openapi-docs/src/test/resources/expected_coproduct.yml]: I checked it using openapi online editor and it didn't complain

**Pull request merged**